### ISSUE TITLE
How to use the storj-up binary inside of a container to retrieve the credentials?

### ISSUE THREAD

**Issue by ifraixedes** [storj/up#45]
With the [v1 release](https://github.com/storj/up/releases/tag/v1.0.0) `storj-up` binary is present inside the containers, however, the `storj-up credentials` command doesn't seem to work.

That's the only command that [the integration tests need](https://github.com/storj-thirdparty/uplink-rust/blob/d9947193e37f819f4c46d4304b701b6c49622bdf/uplink/Makefile#L42).

The error is the following 

```
docker compose exec satellite-api storj-up credentials -e
#Server is not yet available. Retry in 1 sec... rpc: dial tcp 127.0.0.1:7777: connect: connection refused
#Server is not yet available. Retry in 1 sec... rpc: dial tcp 127.0.0.1:7777: connect: connection refused
#Server is not yet available. Retry in 1 sec... rpc: dial tcp 127.0.0.1:7777: connect: connection refused
#Server is not yet available. Retry in 1 sec... rpc: dial tcp 127.0.0.1:7777: connect: connection refused
#Server is not yet available. Retry in 1 sec... rpc: dial tcp 127.0.0.1:7777: connect: connection refused
#Server is not yet available. Retry in 1 sec... rpc: dial tcp 127.0.0.1:7777: connect: connection refused
#Server is not yet available. Retry in 1 sec... rpc: dial tcp 127.0.0.1:7777: connect: connection refused
#Server is not yet available. Retry in 1 sec... rpc: dial tcp 127.0.0.1:7777: connect: connection refused
#Server is not yet available. Retry in 1 sec... rpc: dial tcp 127.0.0.1:7777: connect: connection refused
#Server is not yet available. Retry in 1 sec... rpc: dial tcp 127.0.0.1:7777: connect: connection refused
#Server is not yet available. Retry in 1 sec... rpc: dial tcp 127.0.0.1:7777: connect: connection refused
```

If I'm not mistaken that's because inside a container the other containers run in the docker subnetwork and they are not mapped to the container localhost.

I'm looking for a workaround or to know if the implementation can be changed in a way that the command work when it runs inside of a container.

**Comment by elek** [comment]
You can see `test/test-edge.sh` as an example (fixed on `v2` branch):

```
eval $(docker-compose exec -T satellite-api storj-up credentials --s3 -e -a http://authservice:8888 -s satellite-api)
```

TLDR: use -`s` and `-a`

This generates an access grant with executing the *storj-up credentials* *inside* the container.

Hopefully it will be merged to main, soon...

**Comment by ifraixedes** [comment]
OK cool, however, I won't use it in https://github.com/storj-thirdparty/uplink-rust/issues/33 until it gets released.

I leave the ticket open until v2 is published.

**Comment by ifraixedes** [comment]
Oh, I'm realizing that you posted the workaround for v1. The v2 mention confused me to understand that you meant that it will be fixed in v2.

Thank you.

**Comment by ifraixedes** [comment]
:thinking: now I have another issue.

Using the flag `-s satellite-api` generates an access grant whose us `satellite-api` as satellite's address

```
$ uplink access inspect 1Boks7vVdKd6CYAaa6S4xvpUYAhq2y67w3pXwWit6dRyuxivM3f6J5NWbJFJm3JPbW9mACcZ8KfzFSazo4345s8Fuz7exkRVSzFHUodVbL2gHCPgHqHtV22siminGF3sB8VvfY2eX7sB9BWnfCq8wnT8j59RK4N2VZSWR9QHNousaALCXSB2WJB26zgwPE91zJinYDvtw8ULhQcpN1WGyZWgHhw3r7tQqg1HEy4es4peP6zD1xaLfHyE74VCa9hV

{
  "satellite_addr": "12whfK1EDvHJtajBiAUeajQLYcWqxcQmdYQU5zX5cCf6bAxfgu4@satellite-api:7777",
  "encryption_access": {
    "default_key": "SJo39j/iCMCSK2fe1K5X7Xxj70Y11CIjpfhNsInf624=",
    "default_path_cipher": "ENC_AESGCM"
  },
  "api_key": "13Yqg9xRxBKqPQ3wpomPaFuh6ZLr56D3K3HuMvQZTcwgstpu28SZ6N2Eh6AZjp97a4Y5LVod6NZsps5QUm5WpvUrgZmFd6sWzQD2GFE",
  "macaroon": {
    "head": "3LBqU-DtOWaOdVDwVFFQ2LJSasNf7igpXEU5TjemLFE=",
    "caveats": [],
    "tail": "XYy96emOYQrEiuSWuyjpbcVHkAMmI72linYPGddeL38="
  }
}
```

This address doesn't resolve from the host machine so the tests fail.

**Comment by elek** [comment]
Well, the idea is if you would like to use sg. from host machine, you should use storj-up from the host machine... If you would like to use the docker env, you should use the docker-env.

Looks to be more simple to `go install ...storj-up` on host machine (or download the release)

Note: after `storj/storj@7f6f7e124` you can use `STORJ_DATABASE_OPTIONS_MIGRATION_UNSAFE=snapshot,testdata`. It will create a user (and api key) with the same ID every time. It would allow to use the same grant all the time without using any `storj-up credentials` command.

Unfortunately we need a new release...

**Comment by ifraixedes** [comment]
> Looks to be more simple to go install ...storj-up on host machine (or download the release)

This is what currently the Rust bindings are doing. I was thinking to get rid of it to reduce CI time, but it's OK to leave it as it's.

> Note: after storj/storj@7f6f7e124 you can use STORJ_DATABASE_OPTIONS_MIGRATION_UNSAFE=snapshot,testdata. It will create a user (and api key) with the same ID every time. It would allow to use the same grant all the time without using any storj-up credentials command.
>
> Unfortunately we need a new release...

This could be another solution to reduce the complexity of integration tests external dependencies and CI because storj-up isn't needed and we can hard code all that environment data in the test sources.

---

Let me ask you one thing, I had to do this workaround https://github.com/storj-thirdparty/uplink-rust/pull/35/files#diff-f030db3970e3a1ee586095c81696c003a52a3f1ffe2e071967173dc8c568977bR39 for using v1.0.0

Right now, I cannot say if doing that will work for testing the uplink API https://pkg.go.dev/storj.io/uplink@v1.9.0/edge#Config.RegisterAccess because I haven't started to work on those.

Do you know if the values that I get will work if the tests are executed from the host machine?

If you don't know it's fine, I'll figure it out when I work on them.

Feel free to close the issue because I think that there isn't anything to follow here rather than waiting for a new release, but I don't believe that this issue should be open until then because the solutions are already mentioned.


Last but not least,  thank you for your time!

**Comment by elek** [comment]
> Do you know if the values that I get will work if the tests are executed from the host machine?

Don't think so, unfortunately, but it's just a guess. Try it out, and will see ;-)  I think it will generate grant which can be used only from containers.

One more comment:

With `v2` we have an interface to generate different environments. Which opens the possibility to create `docker-compose cluster` with host networking easily. Which means that you can do this without the pain of tracking being outside/inside...
